buffer/src/main/java/io/netty/buffer/api/DefaultCompositeBuffer.java

```diff
@@ -447,6 +453,83 @@ public void copyInto(int srcPos, Buffer dest, int destPos, int length) {
         }
     }
 
+    @Override
+    public int readIntoChannelWrite(WritableByteChannel channel, int length) throws IOException {
+        if (!isAccessible()) {
+            throw bufferIsClosed(this);
+        }
+        length = Math.min(readableBytes(), length);
+        checkReadBounds(readerOffset(), length);
+        ByteBufferCollector collector = new ByteBufferCollector(countReadableComponents());
+        forEachReadable(0, collector);
+        ByteBuffer[] byteBuffers = collector.buffers;
+        int bufferCount = countAndPrepareBuffersForChannelIO(length, byteBuffers);
+        int totalBytesWritten = 0;
+        if (channel instanceof GatheringByteChannel) {
+            GatheringByteChannel gatheringChannel = (GatheringByteChannel) channel;
+            totalBytesWritten = toIntExact(gatheringChannel.write(byteBuffers, 0, bufferCount));
+            skipReadable(totalBytesWritten);
+        } else {
+            for (int i = 0; i < bufferCount; i++) {
+                int bytesWritten = channel.write(byteBuffers[i]);
+                skipReadable(bytesWritten);
+                totalBytesWritten = addExact(totalBytesWritten, bytesWritten);
+            }
+        }
```

Comment:
```suggestion
        if (channel instanceof GatheringByteChannel) {
            GatheringByteChannel gatheringChannel = (GatheringByteChannel) channel;
            totalBytesWritten = toIntExact(gatheringChannel.write(byteBuffers, 0, bufferCount));
        } else {
            for (int i = 0; i < bufferCount; i++) {
                int bytesWritten = channel.write(byteBuffers[i]);
                totalBytesWritten = addExact(totalBytesWritten, bytesWritten);
            }
        }
        skipReadable(totalBytesWritten);
```

Reply:
We need to also handle if a not-first `channel.write(byteBuffers[i])` call throws. Buffer and channel positions needs to move together for partial failures.

Reply:
we could still do this by just do it in a finally block tho